### PR TITLE
Implement faster thread local rng for scheduler

### PR DESCRIPTION
**Pull request by gbaraldi**
Implement optimal uniform random number generator using the method proposed in https://github.com/swiftlang/swift/pull/39143 based on OpenSSL's implementation of it in https://github.com/openssl/openssl/blob/1d2cbd9b5a126189d5e9bc78a3bdb9709427d02b/crypto/rand/rand_uniform.c#L13-L99

This PR also fixes some bugs found while developing it. This is a replacement for https://github.com/JuliaLang/julia/pull/50203 and fixes the issues found by @IanButterworth with both rngs 

C rng
image

New scheduler rng
image

~On my benchmarks the julia implementation seems to be almost 50% faster than the current implementation.~
With oscars suggestion of removing the debiasing this is now almost 5x faster than the original implementation. And almost fully branchless 

We might want to backport the two previous commits since they technically fix bugs.

### CODE REVIEW

File: base/partr.jl [hunk 1]
```diff
@@ -19,9 +19,95 @@ const heap_d = UInt32(8)
...
+# it overflows, the carry propagates to the integer part (return i+1).
+# If it can no longer overflow regardless of further lower order bits,
+# we are done (return i).  If there is still a chance of overflow, we
+# repeat the process with the next lower word.
+#
+# Each *bit* of randomness has a probability of one half of terminating
+# this process, so each each word beyond the first has a probability
+# of 2^-32 of not terminating the process.  That is, we're extremely
+# likely to stop very rapidly.
+    for _ in 1:10
```
Comment (oscardssmith) [inline base/partr.jl:93]: I feel like we should be able to delete this loop entirely. Without it, we have a bias of 2^-32 which seems like it should be plenty low for the purposes of guaranteeing uniform scheduling

Comment (gbaraldi) [reply base/partr.jl:95]: That's a fair point
image
This seems uniform enough.

**Comment by gbaraldi** [comment]
Hmm, this seems to have made i686 very unhappy

**Comment by giordano** [comment]
> We might want to backport the two previous commits since they technically fix bugs.

Sounds like that should be a separate PR

**Comment by IanButterworth** [comment]
Potentially naive question, but do we have a good benchmark to demonstrate that a faster scheduler rng actually speeds up scheduling?

I ask because in my testing of https://github.com/JuliaLang/julia/pull/50203 I saw the surprising behavior that the threaded fib got slower the faster the rng used (with no changes to allocations) 
```
function fib(n::Int)
    n < 2 && return n
    t = Threads.@spawn fib(n - 2)
    return fib(n - 1) + fetch(t)
end
```

**Comment by gbaraldi** [comment]
I don't think it will affect it too too much. Because if the rand call is super hot then things aren't going so well elsewhere :)

File: base/partr.jl [hunk 2]
```diff
@@ -20,7 +20,56 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
 const heaps_lock = [SpinLock(), SpinLock()]
 
 
-cong(max::UInt32) = iszero(max) ? UInt32(0) : ccall(:jl_rand_ptls, UInt32, (UInt32,), max) + UInt32(1)
```
Comment (giordano) [inline base/partr.jl:23]: Should https://github.com/JuliaLang/julia/blob/0a26e908e77f147a5a964735ad02f74d047fe47b/src/scheduler.c#L90-L94 be deleted, as suggested by https://github.com/JuliaLang/julia/pull/50203/files#diff-7ba7e0372b2249c6b36e693fccd2daca24190ef3001e6e61ac625c02b4bae4d2R91?

File: base/partr.jl [hunk 3]
```diff
@@ -20,7 +20,57 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
...
+get_ptls_rng() = ccall(:jl_get_ptls_rng, UInt64, ())
+
+set_ptls_rng(seed::UInt64) = ccall(:jl_set_ptls_rng, Cvoid, (UInt64,), seed)
+
+"""
+    jl_rand_ptls(max::UInt32)
+Return a random UInt32 in the range `0:max-1` using the thread-local RNG
+state. Max must be greater than 0.
+"""
+Base.@assume_effects :removable :inaccessiblememonly :notaskstate function jl_rand_ptls(max::UInt32)
```
Comment (vchuravy) [inline base/partr.jl:38]:
```suggestion
Base.@assume_effects :removable :inaccessiblememonly :notaskstate function rand_ptls(max::UInt32)
```

File: base/partr.jl [hunk 4]
```diff
@@ -20,7 +20,57 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
 const heaps_lock = [SpinLock(), SpinLock()]
 
 
-cong(max::UInt32) = iszero(max) ? UInt32(0) : ccall(:jl_rand_ptls, UInt32, (UInt32,), max) + UInt32(1)
+"""
+    cong(max::UInt32)
```
Comment (vchuravy) [inline base/partr.jl:24]:
```suggestion
    cong(max::UInt32)

```

File: base/partr.jl [hunk 5]
```diff
@@ -20,7 +20,57 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
...
+Return a random UInt32 in the range `1:max` except if max is 0, in that case return 0.
+"""
+cong(max::UInt32) = iszero(max) ? UInt32(0) : jl_rand_ptls(max) + UInt32(1) #TODO: make sure users don't use 0 and remove this check
+
+get_ptls_rng() = ccall(:jl_get_ptls_rng, UInt64, ())
+
+set_ptls_rng(seed::UInt64) = ccall(:jl_set_ptls_rng, Cvoid, (UInt64,), seed)
+
+"""
+    jl_rand_ptls(max::UInt32)
```
Comment (vchuravy) [inline base/partr.jl:34]:
```suggestion
    rand_ptls(max::UInt32)

```

File: base/partr.jl [hunk 6]
```diff
@@ -20,7 +20,57 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
 const heaps_lock = [SpinLock(), SpinLock()]
 
 
-cong(max::UInt32) = iszero(max) ? UInt32(0) : ccall(:jl_rand_ptls, UInt32, (UInt32,), max) + UInt32(1)
+"""
+    cong(max::UInt32)
+Return a random UInt32 in the range `1:max` except if max is 0, in that case return 0.
+"""
+cong(max::UInt32) = iszero(max) ? UInt32(0) : jl_rand_ptls(max) + UInt32(1) #TODO: make sure users don't use 0 and remove this check
```
Comment (vchuravy) [inline base/partr.jl:27]:
```suggestion
cong(max::UInt32) = iszero(max) ? UInt32(0) : rand_ptls(max) + UInt32(1) #TODO: make sure users don't use 0 and remove this check
```

File: base/partr.jl [hunk 7]
```diff
@@ -20,7 +20,57 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
...
+# Comments are vendored from their implementation as well.
+# For the original developer check the PR to swift https://github.com/apple/swift/pull/39143.
+
+# Essentially it boils down to incrementally generating a fixed point
+# number on the interval [0, 1) and multiplying this number by the upper
+# range limit.  Once it is certain what the fractional part contributes to
+# the integral part of the product, the algorithm has produced a definitive
+# result.
+"""
+    rand_uniform_max_int32(max::UInt32, seed::UInt64)
```
Comment (vchuravy) [inline base/partr.jl:56]:
```suggestion
    rand_uniform_max_int32(max::UInt32, seed::UInt64)

```

File: base/partr.jl [hunk 8]
```diff
@@ -20,7 +20,57 @@ const heaps = [Vector{taskheap}(undef, 0), Vector{taskheap}(undef, 0)]
...
+Max must be greater than 0.
+"""
+Base.@assume_effects :total function rand_uniform_max_int32(max::UInt32, seed::UInt64)
+    if max == UInt32(1)
+        return UInt32(0), seed
+    end
+# We are generating a fixed point number on the interval [0, 1).
+# Multiplying this by the range gives us a number on [0, upper).
+# The high word of the multiplication result represents the integral part
+# This is not completely unbiased as it's missing the fractional part of the original implementation but it's good enough for our purposes
```
Comment (vchuravy) [inline base/partr.jl:67]:
```suggestion
    # We are generating a fixed point number on the interval [0, 1).
    # Multiplying this by the range gives us a number on [0, upper).
    # The high word of the multiplication result represents the integral part
    # This is not completely unbiased as it's missing the fractional part of the original implementation but it's good enough for our purposes
```

**Review by vchuravy — APPROVED** [review]

File: base/partr.jl (hunk 2 above)
Comment (giordano) [reply base/partr.jl:23]: Bump

**Comment by giordano** [comment]
There are already merge conflicts

File: src/scheduler.c [hunk 9]
```diff
@@ -84,15 +84,6 @@ JL_DLLEXPORT int jl_set_task_threadpoolid(jl_task_t *task, int8_t tpid) JL_NOTSA
 extern int jl_gc_mark_queue_obj_explicit(jl_gc_mark_cache_t *gc_cache,
                                          jl_gc_markqueue_t *mq, jl_value_t *obj) JL_NOTSAFEPOINT;
 
-// parallel task runtime
-// ---
-
-JL_DLLEXPORT uint32_t jl_rand_ptls(uint32_t max) // [0, n)
-{
-    jl_ptls_t ptls = jl_current_task->ptls;
-    return cong(max, &ptls->rngseed);
```
Comment (vchuravy) [inline src/scheduler.c:93]:
```
src/threading.c
344:    return jl_current_task->ptls->rngseed;
350:    jl_current_task->ptls->rngseed = new_seed;
366:    ptls->rngseed = jl_rand();

src/gc-stock.h
449:    int v = gc_first_tid + (int)cong(jl_n_markthreads, &ptls->rngseed); // cong is [0, n)
```

So `ptls->rngseed` is still directly used by the GC.

Comment (gbaraldi) [reply src/scheduler.c:93]: Right. But is that an issue?

Comment (vchuravy) [reply src/scheduler.c:93]: It's more that we have two slightly different algorithms use the same state variable, I was looking at where `cong` was being used

Comment (gbaraldi) [reply src/scheduler.c:93]: They use the same rng update if that's any concern

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/55501#issuecomment-2318921742) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/a95aca9_vs_da3468c/report.html) is available.

**Comment by giordano** [comment]
Do we need to run benchmarks, too?

**Comment by gbaraldi** [comment]
I don't think any benchmark is gonna show this. At least no current one. Calling the function is 5x faster that's all I can say

**Pull request merged**